### DISCUSSION TITLE
Cannot update a record due to "Unique constraint failed"

### DISCUSSION THREAD

**Discussion by Deliaz** [prisma/prisma#19022] [Q&A] [answered]
I have the following prisma schema (Postgres)

```prisma
model Member {
  id             String           @id @default(uuid())
  token          String           @unique @default(uuid())
  summary        Summary[]
}

model Summary {
  id           String   @id @default(uuid())
  date         DateTime @db.Date
  activityTime Int
  memberToken  String
  member       Member   @relation(fields: [memberToken], references: [id])

  @@unique([date, memberToken])
}
```
I am trying to update a record of Summary or create if it doesn't exist:

```js
const summary = await prismadb.summary.findUnique({
  where: {
    date_memberToken: {
      date,
      memberToken: payload.token,
    },
  },
});

console.log({summary}); // returns null even if the record already created

if (summary) {
  await prismadb.summary.update({
    where: {
      id: summary.id,
    },
    data: {
      activityTime: {
        increment: timeSpentInc,
      },
    },
  });
} else {
  await prismadb.summary.create({
    data: {
      date,
      activityTime: timeSpentInc,
      member: {
        connect: {
          token: payload.token,
        },
      },
    },
  });
}
```

It creates a record first time, but fails when I try to update it:

```
"Invalid `prisma.summary.create()` invocation: 
Unique constraint failed on the fields: (`date`,`memberToken`)"
```

It also can't find the record that was just created.
I tried via upsert as well, but the result is the same:

```js
await prismadb.summary.upsert({
  where: {
    date_memberToken: {
      memberToken: payload.token,
      date,
    },
  },
  update: {
    activityTime: {
      increment: timeSpentInc,
    },
  },
  create: {
    date,
    activityTime: timeSpentInc,
    member: {
      connect: {
        token: payload.token,
      },
    },
  },
});
```

What is the mistake here?

**Comment by Deliaz** [comment]
The weird part is that I tried to rename the relation via 
```
@@unique([date, memberToken], name: "date_token")
```
But it still complains about `Summary_memberToken_fkey (index)`

**Comment by Ayush-v** [comment]
Maybe This is causing the `findUnique` and `upsert` queries to fail, as they are looking for a unique constraint with the name `date_memberToken`, which does not exist. To fix the issue, you should change the query to use the correct name for the unique constraint:

```
where: {
  date_memberToken: {
    date: date,
    member: {
      token: payload.token,
    },
  },
},
```

Hopefully with this change, the query will work correctly.

**Reply by Ayush-v** [reply, marked as answer]
Else The issue here could be the `@@unique` constraint in the `Summary` model is not being properly enforced. The unique constraint is defined on the combination of the `date` and `memberToken` fields, but in your code, you are only querying based on those fields separately. 

To properly query based on the unique constraint, you need to concatenate the `date` and `memberToken` fields together using the Prisma `AND` function. You can do this by updating your `where` clause in your `findUnique` and `upsert` queries like this:

```
where: {
  date_memberToken: {
    AND: [date, payload.token]
  }
}
```

This will ensure that the query is properly looking for a unique record based on the combination of `date` and `memberToken`.

**Reply by ludralph** [reply]
Hey @Deliaz 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by ishaangandhi** [reply]
So you have to use `AND`? The documentation for compound indexes doesn't mention that. In fact, it explicitly says you can use the query approach @Deliaz uses.

https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/working-with-composite-ids-and-constraints#updating-and-upserting-records-by-a-compound-id-or-unique-constraint

**Comment by Deliaz** [comment]
Yes, thank you. The question can be closed
